### PR TITLE
storage: Fix unexpected flow control after unsafe destroy range

### PR DESCRIPTION
**Pull request by Connor1996**
### What is changed and how it works?

Issue Number: Close https://github.com/tikv/tikv/issues/17304

What's Changed:

```commit-message
Fix unexpected flow control after unsafe destroy range

Flow controller detects pending compaction bytes jump before and after unsafe destroy range. If there is a jump, the controller enters a state that would ignore the high pending compaction bytes until it falls back to normal. Previously, the controller may not enter the state if the pending compaction bytes is lower than the threshold while long term average pending bytes is still high. Then it would trigger flow control mistakenly.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [x] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
Fix the possible flow control after dropping a large table or partition
```

### CODE REVIEW

**Review by LykxSassinator — APPROVED** [review]
Overall LGTM

File: src/storage/txn/flow_controller/singleton_flow_controller.rs [hunk 1]
```diff
@@ -593,14 +593,24 @@ impl<E: FlowControlFactorStore + Send + 'static> FlowChecker<E> {
                 if !current_cfg.enable {
                     return;
                 }
+                if self.wait_for_destroy_range_finish {
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/storage/txn/flow_controller/singleton_flow_controller.rs:596]: concurrent destroys should be rare, so just keep it simple

File: src/storage/txn/flow_controller/singleton_flow_controller.rs [hunk 2]
```diff
@@ -625,9 +635,13 @@ impl<E: FlowControlFactorStore + Send + 'static> FlowChecker<E> {
                             SCHED_THROTTLE_ACTION_COUNTER
                                 .with_label_values(&[cf, "pending_bytes_jump"])
                                 .inc();
-                        } else {
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply src/storage/txn/flow_controller/singleton_flow_controller.rs:628]: If that's the case, the pending bytes probably have already exceeded the soft limit before unsafe destroy range. Indeed, there is still a chance, but it's hard to recognize the corner case by code, so just ignore it.

[the remark this answers is not in the document]
Comment (glorv) [reply src/storage/txn/flow_controller/singleton_flow_controller.rs:628]: Please add a comment for not cleaning the `pending_bytes_before_unsafe_destroy_range`

**Review by glorv — APPROVED** [review]

**Pull request merged**